### PR TITLE
Feature/355 - Add strongly typed events

### PR DESCRIPTION
**Pull request by samwoodall**
## Description
- Add event sealed class with class for each of the event types
- Remove LocalBroadcastManager support;
- Add `addListener` method to EventEmitter that accepts an interface with a method for each event type.  eg. `fun ExperiencePresented(val experience: Experience)`.

#355

### CODE REVIEW

File: sdk/src/main/kotlin/io/rover/sdk/data/events/AnalyticsService.kt [hunk 1]
```diff
@@ -56,32 +55,34 @@ class AnalyticsService(
...
 
-    private fun sendRequest(eventInformation: EventEmitter.Event) {
+    private fun sendRequest(eventInformation: RoverEvent) {
         val urlRequest = buildRequest(URL(ANALYTICS_ENDPOINT), accountToken)
         val bodyData = encodeBody(eventInformation)
 
         request(urlRequest, bodyData)
     }
 
+    private fun flattenEvent(roverEvent: RoverEvent) = mapOf<String, Any>()
```
Comment (orospakr) [inline sdk/src/main/kotlin/io/rover/sdk/data/events/AnalyticsService.kt:84]: This looks like it's just dropping all the contents of the RoverEvent?

File: sdk/src/main/kotlin/io/rover/sdk/data/events/RoverEvent.kt [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import org.json.JSONObject
+
+sealed class RoverEvent {
+    abstract fun encodeJson(): JSONObject
+
+    data class BlockTapped(val experience: Experience,
+                           val screen: Screen,
+                           val block: Block,
+                           val row: Row,
+                           val campaignId: String?) : RoverEvent() {
```
Comment (orospakr) [inline sdk/src/main/kotlin/io/rover/sdk/data/events/RoverEvent.kt]: Current source formatting style is indents of 4, not matching opening brace. I think this is the most widely accepted Kotlin code style. See https://ktlint.github.io

File: sdk/src/main/kotlin/io/rover/sdk/data/operations/data/Experience.kt [hunk 3]
```diff
@@ -240,12 +240,12 @@ internal fun VerticalAlignment.encodeJson(): JSONObject {
...
+                putProp(this@encodeJson, VerticalAlignment.Stacked::height, "height") { it.encodeJson() }
                 putProp(this@encodeJson, VerticalAlignment.Stacked::topOffset, "topOffset")
                 "VerticalAlignmentStacked"
             }
             is VerticalAlignment.Top -> {
-                putProp(this@encodeJson, VerticalAlignment.Top::height, "height")
+                putProp(this@encodeJson, VerticalAlignment.Top::height, "height") { it.encodeJson() }
                 putProp(this@encodeJson, VerticalAlignment.Top::offset, "bottomOffset")
                 "VerticalAlignmentTop"
             }
```
Comment (orospakr) [inline sdk/src/main/kotlin/io/rover/sdk/data/operations/data/Experience.kt:251]: lol. whoops 😅

Comment (orospakr) [reply sdk/src/main/kotlin/io/rover/sdk/data/operations/data/Experience.kt:251]: wait. how was the end-to-end comprehensive_experience.json test passing before??

**Review by orospakr — CHANGES_REQUESTED** [review]

File: sdk/src/main/kotlin/io/rover/sdk/services/EventEmitter.kt [hunk 4]
```diff
@@ -17,43 +13,51 @@ import org.reactivestreams.Publisher
  * Alternatively, this service emits them as local Android broadcast intents.  This allows consumers
  * to receive them that cannot otherwise link against types in this library.
```
Comment (orospakr) [inline sdk/src/main/kotlin/io/rover/sdk/services/EventEmitter.kt]: oh this comment needs to be removed

File: sdk/src/main/kotlin/io/rover/sdk/data/events/AnalyticsService.kt (hunk 1 above)
Comment (samwoodall) [reply sdk/src/main/kotlin/io/rover/sdk/data/events/AnalyticsService.kt:84]: It is 😢 This is intentional and will be fixed in https://github.com/RoverPlatform/rover-android/pull/358

**Review by orospakr — APPROVED** [review]

**Pull request merged**